go-controller/pkg/clustermanager/clustermanager.go

```diff
@@ -0,0 +1,275 @@
+	rl, err := resourcelock.New(
+		resourcelock.LeasesResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		"ovn-kubernetes-cluster-manager",
+		cm.client.CoreV1(),
+		cm.client.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      cm.identity,
+			EventRecorder: cm.recorder,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	lec := leaderelection.LeaderElectionConfig{
+		Lock:            rl,
+		LeaseDuration:   time.Duration(config.ClusterMgrHA.ElectionLeaseDuration) * time.Second,
+		RenewDeadline:   time.Duration(config.ClusterMgrHA.ElectionRenewDeadline) * time.Second,
+		RetryPeriod:     time.Duration(config.ClusterMgrHA.ElectionRetryPeriod) * time.Second,
+		ReleaseOnCancel: true,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				klog.Infof("Won leader election; in active mode")
+				// run the cluster controller to init the cluster manager
+				start := time.Now()
+				defer func() {
+					end := time.Since(start)
+					metrics.MetricClusterManagerReadyDuration.Set(end.Seconds())
+				}()
+
+				// run only on the active master node.
+				if err := cm.Run(); err != nil {
+					panic(err.Error())
+				}
+			},
+			OnStoppedLeading: func() {
+				// This node was leader and it lost the election.
+				// Whenever the node transitions from leader to follower,
+				// we need to handle the transition properly like clearing
+				// the cache. It is better to exit for now.
+				// kube will restart and this will become a follower.
+				klog.Infof("No longer leader; exiting")
+				os.Exit(0)
+			},
+			OnNewLeader: func(newLeaderName string) {
+				if newLeaderName != cm.identity {
+					klog.Infof("Lost the election to %s; in standby mode", newLeaderName)
+				}
+			},
+		},
+	}
+
+	leaderelection.SetProvider(ovnkubeClusterManagerLeaderMetricsProvider{})
+	leaderElector, err := leaderelection.NewLeaderElector(lec)
+	if err != nil {
+		return err
+	}
+
+	cm.wg.Add(1)
+	go func() {
+		leaderElector.Run(ctx)
+		klog.Infof("Stopped leader election")
+		cm.wg.Done()
+	}()
+
+	return nil
+}
+
+// Run starts the watch factory, inits and runs the default network cluster controller
+func (cm *ClusterManager) Run() error {
+	metrics.RegisterClusterManagerFunctional()
+	if err := cm.defaultNetClusterController.Init(); err != nil {
+		return err
+	}
+
+	if cm.secondaryNetClusterControllerManager != nil {
+		if err := cm.secondaryNetClusterControllerManager.Start(); err != nil {
+			return err
+		}
+	}
+
+	// Start and sync the watch factory to begin listening for events
+	if err := cm.wf.Start(); err != nil {
+		return err
+	}
+
+	if err := cm.defaultNetClusterController.Run(); err != nil {
+		return err
+	}
+
+	if cm.secondaryNetClusterControllerManager != nil {
+		klog.Infof("Starting multi network attach manager")
+		return cm.secondaryNetClusterControllerManager.Run(cm.stopChan)
```

Comment:
As I mentioned earlier,  I followed the existing calling pattern from here - 

https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/network-controller-manager/network_controller_manager.go#L401

I guess this can be done as part of the cleanup PR ? which we discussed above ?
i.e a separate PR for proper usage of  stopChannel, waitgroup ?

Reply:
It is not the same pattern. 

There `cm.nadController.Run(cm.stopChan)` is done like that because the `nadController` has no Stop method.

Whereas here `secondaryNetClusterControllerManager` has a Stop method.